pkg/pillar/utils/deserialize.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2017-2023 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package utils
+
+import (
+	"encoding/json"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"os"
+	"reflect"
+)
+
+func readFile(filename string) ([]byte, error) {
+	// Check if the file exists
+	if _, err := os.Stat(filename); os.IsNotExist(err) {
+		log.Errorf("File %s does not exist", filename)
+		return nil, err
+	}
+
+	// Open the file
+	file, err := os.Open(filename)
+	if err != nil {
+		log.Errorf("Failed to open file %s, %s", filename, err.Error())
+		return nil, err
+	}
+	defer file.Close()
+
+	// Read the data
+	data, err := io.ReadAll(file)
+	if err != nil {
+		log.Errorf("Failed to read data from file %s, %s", filename, err.Error())
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func validateFields(dataMap map[string]interface{}, expectedFields map[string]bool, criticalFields map[string]bool, filename string) error {
+	for k := range dataMap {
+		if _, ok := expectedFields[k]; !ok {
+			log.Warnf("Unexpected field %s in stored data in %s", k, filename)
+		}
+	}
+
+	for k := range expectedFields {
+		if _, ok := dataMap[k]; !ok {
+			if criticalFields[k] {
+				errMsg := fmt.Sprintf("Critical field %s missing in stored data in %s", k, filename)
+				log.Errorf(errMsg)
+				return fmt.Errorf(errMsg)
+			}
+			log.Warnf("Missing field %s in stored data in %s", k, filename)
+		}
+	}
+
+	return nil
+}
+
+func extractFields(t reflect.Type, fieldsMap map[string]bool, criticalFields map[string]bool) {
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+
+		if field.Anonymous {
+			// If this is an anonymous field, recursively extract its fields as well.
+			extractFields(field.Type, fieldsMap, criticalFields)
+		} else {
+			// This is a normal field, so just add its name to the map.
+			fieldsMap[field.Name] = true
+			// Check if the field is mandatory
+			tag := field.Tag.Get("mandatory")
+			if tag == "true" {
+				criticalFields[field.Name] = true
+			}
+		}
+	}
+}
+
+// DeserializeToStruct deserializes the given file into the given struct.
+func DeserializeToStruct(filename string, pointerToStructInstance any) error {
```

Comment:
Nit, but it wouldn't hurt to comment that this will check and fail if any fields with mandatory=true are not present.

Reply:
Done. 
I also added a check that the pointer passed to the function is a pointer to a struct.